experimental/report/renderer.go

```diff
@@ -48,34 +49,42 @@ type Renderer struct {
 
 // Render renders a diagnostic report.
 //
-// In addition to returning the rendering result, returns whether any
-// errors occurred.
-func (r Renderer) Render(report *Report) (text string, haveErrors bool) {
-	var out strings.Builder
-	var errors, warnings int
+// In addition to returning the rendering result, returns whether the report
+// contains any errors.
+//
+// On the other hand, the actual error-typed return is an error when writing to
+// the writer.
+func (r Renderer) Render(report *Report, out io.Writer) (errorCount, warningCount int, err error) {
 	for _, diagnostic := range report.Diagnostics {
 		if !r.ShowRemarks && diagnostic.Level == Remark {
 			continue
 		}
 
-		out.WriteString(r.Diagnostic(diagnostic))
-		out.WriteString("\n")
-		if !r.Austere {
-			out.WriteString("\n")
+		_, err = fmt.Fprintln(out, r.Diagnostic(diagnostic))
+		if err != nil {
```

Comment:
You can save a small amount of vertical real estate by collapsing these, here and throughout
```suggestion
		if _, err = fmt.Fprintln(out, r.Diagnostic(diagnostic)); err != nil {
```